sidekick_core/lib/src/sidekick_version_checker.dart

```diff
@@ -0,0 +1,106 @@
+  /// E.g. ['dev_dependencies', 'some_dev_dependency'] or ['sidekick', 'generator_version']
+  Future<bool> isUpToDate({
+    required String package,
+    List<String>? pubspecPath,
+  }) async {
+    final latest = await getLatestPackageVersion(package);
+    final current = getCurrentMinimumPackageVersion(
+      pubspecPath ?? ['dependencies', package],
+    );
+
+    return latest == current;
+  }
+
+  /// Updates the version constraint of [package] to the latest version
+  Future<void> updateVersionConstraintToLatest(String package) async =>
+      updateVersionConstraint(
+        package: package,
+        newMinimumVersion: await getLatestPackageVersion(package),
+      );
+
+  /// Replaces the current version constraint of [package] in pubspec.yaml with [newMinimumVersion]
+  ///
+  /// If [pinVersion] is true, the new version constraint will only allow [newMinimumVersion]
+  /// Else it will allow a version range from [newMinimumVersion] until the next breaking version
+  void updateVersionConstraint({
+    required String package,
+    required Version newMinimumVersion,
+    bool pinVersion = false,
+  }) {
+    final pubspec = Repository.requiredSidekickPackage.pubspec;
+    final lines = pubspec.readAsLinesSync();
+
+    final newVersionConstraint = pinVersion
+        ? newMinimumVersion.canonicalizedVersion
+        : newMinimumVersion.major > 0
+            ? '^${newMinimumVersion.canonicalizedVersion}'
+            : "'>=${newMinimumVersion.canonicalizedVersion} <1.0.0'";
+
+    final index = lines.indexWhere((it) => it.startsWith('  $package:'));
+    lines[index] = '  $package: $newVersionConstraint';
+
+    pubspec.writeAsStringSync(lines.join('\n'));
+  }
+
+  /// Returns the latest version of [package] available on pub.dev
+  Future<Version> getLatestPackageVersion(String package) async {
+    final response =
+        await get(Uri.parse('https://pub.dev/api/packages/$package'));
+
+    if (response.statusCode != HttpStatus.ok) {
+      throw "Package '$package' not found on pub.dev";
+    }
+
+    final latestVersion =
+        pick(jsonDecode(response.body), 'latest', 'version').asStringOrThrow();
+
+    return Version.parse(latestVersion);
+  }
+
+  /// Returns the current minimum version of [package] being used in the generated sidekick CLI
+  ///
+  /// [pubspecPath] is the path from which to retrieve the version in pubspec.yaml, e.g.
+  /// - ['dependencies', 'sidekick_core']
+  /// - ['dev_dependencies', 'lint']
+  /// - ['sidekick', 'generator_version']
+  Version getCurrentMinimumPackageVersion(List<String> pubspecPath) {
+    final versionConstraint = _readFromPubspecYaml(pubspecPath);
+
+    final versionConstraintRegEx = RegExp(
+      '[\'"\\^<>= ]*(\\d+\\.\\d+\\.\\d+(?:[+-]\\S+)?)',
+    );
+
+    final minVersion = versionConstraintRegEx
+        .allMatches(versionConstraint)
+        .map((e) => e.group(1))
+        .whereNotNull()
+        .first;
+
+    return Version.parse(minVersion);
```

Comment:
I tried doing that but didn't like it because you end up needing something like this:

```dart
final pubspec =PubSpec.fromFile(Repository.requiredSidekickPackage.pubspec.path);
return (pubspec.dependencies['sidekick_core']?.reference as ExternalHostedReference).versionConstraint as VersionRange).min;
```

With that approach I also ran into this error straight from dependency hell:
```
The name 'PubSpec' is defined in the libraries 'package:dcli/src/pubspec/pubspec.dart' and 'package:pubspec2/src/pubspec.dart (via package:pubspec2/pubspec2.dart)'
```